### ISSUE TITLE
Give nodes a stable scene local id to fix scene inheritance issues

### ISSUE THREAD

**Issue by Flavelius** [godotengine/godot-proposals#6291]
### Describe the project you are working on

A game making heavy use of scene inheritance

### Describe the problem or limitation you are having in your project

When having a longer inheritance chain and renaming nodes in a base scene it breaks every inheriting scene that uses that renamed node hierarchically. That node is now virtually missing in all inherited scenes, even though it has just been renamed, and all nodes placed under it become orphans which weakens one of the core engine features by a considerable amount.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Give each node a scene local unique id that is then used to track hierarchical relationships, instead of just using the brittle name (which is by design intended to be changed/modifyable and thus unfit to be used in contexts that need stable identifiers)

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

When a node gets created it's assigned a scene-local unique id. Simple numeric ids will likely work fine (uint/ulong)

This means each node implicitly gets a project unique id in the process (asset/packedscene-uid + local-nodeid).

tscn's then, instead of just storing nodes via their parent path, also store the direct parent node's [scene-uid + local-nodeid combination]. For non-inherited local-only hierarchies the scene-uid part could even be left out.
This id combination could be wrapped via a specialized type if necessary to ease generation/serialization/instance resolution etc (unity for exaxmple does that)

That way inheritance chains will no longer break by default with renamed nodes. 
The engine's scene deserializer might even benefit from this by then only having to compare ids to assemble scenes instead of name-paths, but can use those as fallback, similar to the current UID system for Resources.
And the change to tscns is quite minimal (one additional parent-id field next to a node's relational path).

For scripting purposes this unique id could potentially even be exposed, so for example networked games could make use of this instead of needing to generate their own.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It cannot be worked around in script, it's core functionality

### Is there a reason why this should be core and not an add-on in the asset library?

It is core functionality and improves one of the main flaws with the current scene inheritance system, so benefits everyone and thus should be readily accessible aswell.

**Comment by YuriSizov** [comment]
If the issue that inherited scenes are not notified about a rename, then that should be addressed, and adding a second identifier is not improving things much. Node's path inside of a scene is already unique, so by knowing the before path and the after path it can be updated without any extra fields.

As always with inherited scenes, there are many issues with extending scenes doing modifications that would conflict with future parent changes. And source of pain is not going anywhere.

**Comment by Flavelius** [comment]
The way i understand it is that godot would need to rewrite many files that are affected by one single rename with multilevel inheritance. This in comparison to none with ids which is not only less errorprone and more VCS friendly aswell but also better for performing because there's no need to be rescanning and modifying files in the background while merely editing scene contents (or needing to track all such actions for the moment of saving, adding enormous unnecessary complexity). It would also cover the case of moving (not just renaming) nodes in the hierarchy where children in inherited scenes will keep their parent which is a better alternative than losing them as orphans

**Comment by YuriSizov** [comment]
> This in comparison to none with ids

You can't avoid changing inherited scenes if you want to preserve the structure of the scene. Node paths are the main way to reference nodes, or node names at the very least if those nodes are scene-unique. Not updating them means that the structure of the inherited scene would not reflect the structure it is inheriting.

**Comment by Flavelius** [comment]
> You can't avoid changing inherited scenes if you want to preserve the structure of the scene. 

 This currently and in this scenario happens when opening the inherited scene. Currently nodes are placed at the root, in the Id case they would be moved to their corrected place in the hierarchy.
As for the ids; then those ids are not the same as the ones in unity and don't relate to some comparison (I meant 'similar' in relation to the idea I proposed). They are not the same as runtime instance ids. They are mainly relevant to scene construction in which case it doesn't matter whether packed scenes can be instantiated multiple times, the same can be done in unity.

**Comment by YuriSizov** [comment]
> whereas in the Id case they would be moved to their corrected place in the hierarchy.

Same can be done based on the rename itself, it's just not handled correctly. There is still no need for a new, secondary ID, as the node's path is already sufficient to identify and move it.

**Comment by Flavelius** [comment]
> 

This then comes back to what the initial proposal and my first comment contain (see arguments there), but fundamentally worth to repeat is that node names are meant to change and in that regard are not a good choice to represent cross-file relations.

Are there any actual downsides to improving the current situation this way? It reads like i'm proposing to destroy something that works okay, which it isn't and ofcourse it doesn't fix all issues, but a major pain point. and (U)IDs are already something that godot finally accepted for similar issues as being useful; it's basically an extension of that idea.

**Comment by markdibarry** [comment]
At least how VS and VSCode handles it, renaming only affects the original item, but you can right click and specifically request that it rename by reference. I imagine it wouldn't be ideal for the editor to sift through the project and look for all inherited scenes every time you do a rename, but maybe when you explicitly request it to, it may be more reasonable?

**Comment by Zireael07** [comment]
@markdibarry This veers into being a duplicate of a proposal to add refactoring (project wide renaming) support...

**Comment by markdibarry** [comment]
True. I guess the question is, what does this solve better than that solution which doesn't require a new id to maintain?

**Comment by Flavelius** [comment]
The complication with a dedicated project wide renaming feature is that the requirement of knowledge is inverted. Users need to know that other scenes inherit from the one they're editing (which there is no indication of naturally) to use the refactor option instead of a normal rename. And if it was the default rename action, the other issues mentioned above arise.
By using ids other scenes only need to resolve hierarchy dependencies the moment _they_ require those, not the other way round, etc, as mentioned.

**Comment by Flavelius** [comment]
I edited the proposal to include an example with comments

**Comment by Flavelius** [comment]
Ok, so it actually turned out to be a little different.
But i created a proof of concept that works quite well and can be found here:

https://github.com/Flavelius/godot/tree/node_inheritance

and here's a test project, but a new project should do it aswell:

[TestProject.zip](https://github.com/godotengine/godot-proposals/files/10807323/TestProject.zip)

Technically, it is currently designed like this:

Every node gets an additional scene-local-id field.
When a scene is saved, nodes that belong to the saved scene get their id assigned a unique value among them. 
This value becomes the node's stable identifier (it's retained through sessions), relative to the scene it resides in (!).

The id is persisted for all nodes that have a valid one assigned (ideally all, more on that later).
Additionally, nodes that have a parent have one extra field being assigned that relates to the scene local id of their parent;
With the special case of nodes that have a parent that is not part of their scene (inherited or instantiated scenes).
These nodes are being assigned one more id that identifies the scene local root of the inserted scene.

This relates to what's additionally written into the tscn (can be inspected in the test project linked above):
```
-all nodes: + scene-local-id
-- if they have a parent: + parent-id
--- if parent is from an inserted scene: + parent-root-id 
```

Upon loading a scene, the current process is almost completely kept as it is, but orphans are tracked and attempted to be reconnected at the end. This means failure cases that are not handled are handed off to the current/default orphan handling procedure.

**This change is backwards compatible**, with a small caveat:
for projects in which scenes already exist without ids, warnings will be printed to the console until those scenes are resaved in order of their inheritance (so ids are valid).

For some reason (may be related to caching?), when saving a modified scene that is already open and instanced/inherited in another scene tab,
the node coloring is lost for inherited/instanced nodes (reconnected node parent names become the usual grey, instead of being yellow tinted), reloading the scene from disk corrects this.

For newly created projects there's no issue as scenes that are inherited need to exist on disk, meaning they will already have valid ids by that time.

### What i haven't tested yet:

- [ ] Signals (whether they remain/become or are valid throughout the reconnection process)
- [ ] Scene Unique names (i have put code in place that calls the acquisition method for them, but not sure if the node is already in an invalid state by then)
- Changing/swapping root nodes (through the context menu; although i think it's expected that this is a breaking change anyway)

### What i have tested:
- [x] renaming nodes (this was the main motivation, seems to work without issues)
- [x] moving nodes (nodes can be moved inside their local hierarchy and will retain children in inherited/instanced scenes)
- [x] copy-paste (this doesn't seem to interfere, copied nodes get a new id assigned to them)
- [x] duplicate (also doesn't interfere, same as copy paste)

___
And here's a demonstration using the linked test project:

https://user-images.githubusercontent.com/8841352/220732782-644e3088-1209-4637-b8c7-7a30df12e6d3.mp4
